[libc][math][c23] add c23 floating point fmaximum and fminimum functions.

Jobhdez opened the pull request:
Hello,

This PR addresses this issue https://github.com/llvm/llvm-project/issues/85496.

I have added fminimum and fmaximum functions.

I built it using 

```
$ cmake ../llvm -G Ninja   -DLLVM_ENABLE_PROJECTS="llvm;libc"   -DLLVM_TARGETS_TO_BUILD="X86"   -DCMAKE_BUILD_TYPE=Debug   -DCMAKE_C_COMPILER=clang   -DCMAKE_CXX_COMPILER=clang++

$ ninja libc -j8
```

I ran the tests as follows:
```
ninja libc-math-smoke-tests
```

Please let me know of anything I need to change. I will be available.

Thanks.

Commit:
[libc][math][c23] add c23 floating point fmaximum and fminimum fns

nickdesaulniers commented on libc/config/windows/entrypoints.txt:
```diff
@@ -1,4 +1,4 @@
-set(TARGET_LIBC_ENTRYPOINTS
+fset(TARGET_LIBC_ENTRYPOINTS
```
typo?

Jobhdez replied:
I will fix it

Jobhdez commented:
> Thanks for the patch! Try `git clang-format HEAD~` to appease the linter!

Do I just do just that. I the. Have to amend the commit?

nickdesaulniers commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -58,6 +58,130 @@ LIBC_INLINE T fmax(T x, T y) {
+
+  if (bitx.is_nan()) {
+    return x;
+  } else if (bity.is_nan()) {
```
Please use the form:

```
if (bitx. ...)
  return ...
if (bity. ...)
  return
if (bitx.sign ...)
  return
return x > y ? x : y;
```

Here and below.

Jobhdez replied:
ok I will fix it. is it ok after I make my changes to do 

```
$ git add
$ git commit --amend
$ git push --force
```
So I can keep it at two commits?

lntue replied:
It's ok for you to keep multiple commits, don't need to `--amend`.

nickdesaulniers commented:
> > Thanks for the patch! Try `git clang-format HEAD~` to appease the linter!
> 
> Do I just do just that. Do I Have to amend the commit?

I'd just commit the result and push it so that you have more than one commit.

nickdesaulniers commented:
There was an open bug for this, yeah? If you put `Fixes #<bug number>` without the `<>` in the commit message, github will autoclose that issue when the PR is merged.

Jobhdez commented:
> There was an open bug for this, yeah? If you put `Fixes #<bug number>` without the `<>` in the commit message, github will autoclose that issue when the PR is merged.

It’s for this issue here https://github.com/llvm/llvm-project/issues/85496

Commit:
format code using clang-format

Jobhdez commented:
> Thanks for the patch! Try `git clang-format HEAD~` to appease the linter!

I just did that and the checks failed again.

Jobhdez commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -58,6 +58,130 @@ LIBC_INLINE T fmax(T x, T y) {
+LIBC_INLINE T fmaximum(T x, T y) {
+  FPBits<T> bitx(x), bity(y);
+
+  if (bitx.is_nan()) {
```
I will fix it. thanks

Jobhdez commented on libc/config/windows/entrypoints.txt:
```diff
@@ -1,4 +1,4 @@
-set(TARGET_LIBC_ENTRYPOINTS
+fset(TARGET_LIBC_ENTRYPOINTS
```
I will fix this.

Jobhdez commented on libc/src/math/amdgpu/fmaximum.cpp:
```diff
@@ -0,0 +1,21 @@
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(double, fmaximum, (double x, double y)) {
+  return __builtin_fmaximum(x, y);
```
Let me know. If these need to be removed I will remove them.

lntue replied:
Let's remove all the specialization in `src/math/amdgpu` folder for now.

Jobhdez replied:
Ok, I will do that.

Jobhdez replied:
so I need to delete the fmaximum.cpp and so on functions in this directory?

lntue replied:
Yes.

lntue commented on libc/config/linux/aarch64/entrypoints.txt:
```diff
@@ -363,6 +363,30 @@ set(TARGET_LIBM_ENTRYPOINTS
+    libc.src.math.fminimum_magl
+    libc.src.math.fminimum_mag_num
+    libc.src.math.fminimum_mag_numf
+    libc.src.math.fminimum_mag_numl
```
Add the `f128` functions in `if(LIBC_TYPES_HAS_FLOAT128)` below.  Same for x86_64 and riscv.

Jobhdez replied:
Ok will do!

lntue commented on libc/src/math/amdgpu/fmaximum.cpp:
```diff
@@ -0,0 +1,22 @@
+//===-- Implementation of the fmaximum function for GPU
+//-----------------------===//
```
Make it fit in 1 line, and exact 80-char.

Jobhdez replied:
same with other files in the directory `amdgpu`? or everywhere else?

jhuber6 replied:
Everywhere, the headers should all look like the other ones.

Jobhdez replied:
ok thanks

lntue commented:
You should also try with full build mode: start with a clean build folder:
```
$ cmake ../llvm -GNinja -DLLVM_ENABLE_PROJECTS="llvm;libcl;clang;compiler-rt" -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_BUILD_TYPE=Release -DLLVM_LIBC_INCLUDE_SCUDO=ON -DLLVM_LIBC_FULL_BUILD=ON
$ ninja libc
$ ninja check-libc
```
And you should be able to test them individually:
```
$ ninja libc.test.src.math.smoke.fmaximumf128_test.__unit__
```

lntue commented:
tag @jhuber6 for AMDGPU builtins, building & testing.

Jobhdez commented:
@lntue I will address your reviews; thanks!

arsenm reviewed (commented):
The implementation should just be call llvm.minimum/maximum

michaelrj-google commented:
@arsenm I assume you mean for the GPU implementation? The generic implementation looks correct already since we have to support both clang and GCC.

nickdesaulniers commented:
> > Thanks for the patch! Try `git clang-format HEAD~` to appease the linter!
> 
> I just did that and the checks failed again.

So what's happening is that the comment on the first line of a few files (see [report](https://github.com/llvm/llvm-project/actions/runs/8366631364/job/22907259985)) is too long. When you run the automated formatter, it's wrapping the line (and appears to do so in a way that doesn't reach a fixed point if the formatter is rerun).

Rather than commit changes suggested by the formatter where it's wrapping the first line comment, you should reject those changes (or not commit them; `git commit -p` will bring up an interactive mode where you accept/reject individual hunks of a diff) or just fix them directly to be 80 columns.

lntue commented:
> @arsenm I assume you mean for the GPU implementation? The generic implementation looks correct already since we have to support both clang and GCC.

Probably for this patch, we can remove all the specializations for GPU, and only add them later if they show performance benefits? @jhuber6

Jobhdez commented:
> > > Thanks for the patch! Try `git clang-format HEAD~` to appease the linter!
> > 
> > 
> > I just did that and the checks failed again.
> 
> So what's happening is that the comment on the first line of a few files (see [report](https://github.com/llvm/llvm-project/actions/runs/8366631364/job/22907259985)) is too long. When you run the automated formatter, it's wrapping the line (and appears to do so in a way that doesn't reach a fixed point if the formatter is rerun).
> 
> Rather than commit changes suggested by the formatter where it's wrapping the first line comment, you should reject those changes (or not commit them; `git commit -p` will bring up an interactive mode where you accept/reject individual hunks of a diff) or just fix them directly to be 80 columns.

Thanks. Once I fix the comment do I just run `git clang-format HEAD~`? I have made two commits already. This command will format my first commit right with all the files?

jhuber6 commented:
> > @arsenm I assume you mean for the GPU implementation? The generic implementation looks correct already since we have to support both clang and GCC.
> 
> Probably for this patch, we can remove all the specializations for GPU, and only add them later if they show performance benefits? @jhuber6

The GPU should use `__builtin_fmin` and `__builtin_fmax`. This should be available for f64, f32, and f16. We don't support `f128` on the GPU at all so it can be ignored. I think in general a lot of platforms have hardware for this, so we should use `#if LIBC_HAS_BUILTIN(__builtin_fmin)` where present (Though it may require some work around zero like before.

For now however I think that it's good to have a C implementation to fall back on.

nickdesaulniers commented:
> Thanks. Once I fix the comment do I just run `git clang-format HEAD~`? I have made two commits already. This command will format my first commit right with all the files?

`git clang-format HEAD~<N>` (without the `<>`) will format the last `N` commits.  `N` is optional when `N` == 1.

So if you have 2 commits `git clang-format HEAD~2` would format just the lines touched in the previous 2 commits.

Jobhdez commented:
> > > @arsenm I assume you mean for the GPU implementation? The generic implementation looks correct already since we have to support both clang and GCC.
> > 
> > 
> > Probably for this patch, we can remove all the specializations for GPU, and only add them later if they show performance benefits? @jhuber6
> 
> The GPU should use `__builtin_fmin` and `__builtin_fmax`. This should be available for f64, f32, and f16. We don't support `f128` on the GPU at all so it can be ignored. I think in general a lot of platforms have hardware for this, so we should use `#if LIBC_HAS_BUILTIN(__builtin_fmin)` where present (Though it may require some work around zero like before.
> 
> For now however I think that it's good to have a C implementation to fall back on.

So I can remove the amdgpu stuff right?

Jobhdez commented:
> > Thanks. Once I fix the comment do I just run `git clang-format HEAD~`? I have made two commits already. This command will format my first commit right with all the files?
> 
> `git clang-format HEAD~<N>` (without the `<>`) will format the last `N` commits. `N` is optional when `N` == 1.
> 
> So if you have 2 commits `git clang-format HEAD~2` would format just the lines touched in the previous 2 commits.

thanks

jhuber6 commented:
> So I can remove the amdgpu stuff right?

I don't think most of these will compile. It depends on both `clang` understanding the builtin and the backend being able to emit the intrinsic. See https://godbolt.org/z/daa5fqxK8, you should be able to plug in whatever builtin to check if it works.

michaelrj-google commented:
> So I can remove the amdgpu stuff right?

Yes, we can add them back in a followup patch if needed.

jhuber6 commented:
> > So I can remove the amdgpu stuff right?
> 
> Yes, we can add them back in a followup patch if needed.

This doesn't update the `gpu/entrypoints.txt` so they wouldn't have been compiled either way.

Jobhdez commented:
> > So I can remove the amdgpu stuff right?
> 
> Yes, we can add them back in a followup patch if needed.

Ok.

Jobhdez commented:
My bad, I accidentally closed the pr. But I just reopened it.

But hey you all, I will start addressing your review today but not sure If I will be able to commit something. Thanks. Ill try to get this done asap.

Commit:
address review

Commit:
add fminimum_mag_num to the math/generic dir

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -64,16 +64,14 @@ LIBC_INLINE T fmaximum(T x, T y) {
     return x;
-  } else if (bity.is_nan()) {
+  }
+  if (bity.is_nan()) {
```
Remove the curly braces `{}` around a single statement after `if`.  Same as above and below.

Jobhdez replied:
ok, I will do that. thanks

lntue commented on libc/src/math/nvptx/fmaximumf.cpp:
```diff
@@ -0,0 +1,22 @@
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(float, fmaximumf, (float x, float y)) {
+  return __builtin_fmaximumf(x, y);
```
Let remove all the specializations in `src/math/nvptx` for now.

Jobhdez replied:
Ok. thanks

Commit:
remove specializations

Commit:
remove curly braces

Commit:
format code

Jobhdez commented:
@lntue ok I addressed your review

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -58,6 +58,102 @@ LIBC_INLINE T fmax(T x, T y) {
+LIBC_INLINE T fmaximum_num(T x, T y) {
+  FPBits<T> bitx(x), bity(y);
+
+  if (bitx.is_nan())
```
I'm sorry for a late suggestion, but I just noticed that in section F.10.9.5 in C23 standard about exceptional handling of `f*_num` functions is different from their non-num versions:
- if both are `NaN`, then a quiet NaN is returned.
- if at least one of them is signaling NaN, then `FE_INVALID` is raised.

Do you mind updating `fmaximum_num` and `fminimum_num` implementations here to take care of those cases, and updating `f*_num_test` (4 of them in total) accordingly.

Thanks,

Jobhdez replied:
Yea I’ll get it done tomorrow. 
are the other functions correct? I tried to follow the doc you mentioned on the issue and also looked at some gcc docs.  I’ll double check tomorrow.

lntue replied:
The others look correct to me.

Jobhdez replied:
Here’s the definition I wa
S basing it on 
```
The fmaximum_num function returns the greater of the two values x and y. If one argument is a number and the other is a NaN, even a signaling NaN, the number is returned. Positive zero is treated as greater than negative zero.
```
I took this from here https://www.gnu.org/software/libc/manual/html_node/Misc-FP-Arithmetic.html

Jobhdez replied:
@lntue hello, do changed the code. 

can you please give an idea of how to test the exception? I’m the codebase I have found examples of `EXPECT_FP_EQ_WITH_EXCEPTION`
But not sure how to use it

lntue replied:
Something like `EXPECT_FP_EQ_WITH_EXCEPTION(aNaN, func(aNaN, sNaN), FE_INVALID);` should work.  And you'll neet to test with all combinations of NaNs.

Jobhdez replied:
Is my handling of the case associated with raising the exception correct? Because I used you example and I’m getting `which is 0 to be greater or equal to 4`. I also tried to return a quiet nan but same thing happened

Jobhdez replied:
I figured it out. i found the bug

Commit:
adhere to the C standard

Commit:
format code

Commit:
add fe invalid tests

Commit:
format code

lntue commented on libc/test/src/math/smoke/FMinimumNumTest.h:
```diff
@@ -0,0 +1,100 @@
+  typedef T (*FMinimumNumFunc)(T, T);
+
+  void testNaN(FMinimumNumFunc func) {
+    EXPECT_FP_EQ(inf, func(FPBits::quiet_nan().get_val(), inf));
+    EXPECT_FP_EQ(inf, func(FPBits::signaling_nan().get_val(), inf));
```
`aNaN` and `sNaN` could be used instead of these.  They are just the same.

Jobhdez replied:
aNaN corresponds to quiet nans and sNaN corresponds to signaling nan right?

lntue replied:
Yes, that's correct.

lntue replied:
And the same for `FMaximumNumTest.h`

Jobhdez replied:
what about for FMaximumMagNum?

lntue replied:
Ditto.  For all the tests, just use `aNaN` and `sNaN` consistently, and only use `FPBits::quiet_nan( ...  )` when you need to set something special in the payloads.

lntue commented on libc/test/src/math/smoke/FMinimumNumTest.h:
```diff
@@ -0,0 +1,100 @@
+    EXPECT_FP_EQ(-0.0, func(-0.0, FPBits::signaling_nan().get_val()));
+    EXPECT_FP_EQ(T(-1.2345), func(FPBits::quiet_nan().get_val(), T(-1.2345)));
+    EXPECT_FP_EQ(T(1.2345), func(T(1.2345), FPBits::quiet_nan().get_val()));
+    EXPECT_FP_EQ(T(-1.2345),
```
Change this to `EXPECT_FP_EQ_WITH_EXCEPTION( ... , FE_INVALID);`

Jobhdez replied:
You’re highlighting 4 lines. Which do I change? Do I change all the lines that you highlighted?

lntue replied:
Sorry, I meant that change all the `EXPECT_FP_EQ` to `EXPECT_FP_EQ_WITH_EXCEPTION` if one of the input is signaling nan.

Jobhdez replied:
So do I need to change the logic in BasicOperations? Because in my code I only raise an exception when one argument is a nan while the other is a signaling nan

lntue replied:
Yes, section F.10.9.5 from https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf states that:
- If one input is a number, and the other input is a NaN (quiet or signaling), then return the number.
- If both inputs are NaN (quiet or signaling), then return a quiet NaN.
- If any of the inputs is a signaling NaN, the invalid floating point exception is raised (even if the other input is a number)

Jobhdez replied:
> Yes, section F.10.9.5 from https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf states that:
> 
>     * If one input is a number, and the other input is a NaN (quiet or signaling), then return the number.
> 
>     * If both inputs are NaN (quiet or signaling), then return a quiet NaN.
> 
>     * If any of the inputs is a signaling NaN, the invalid floating point exception is raised (even if the other input is a number)
> 
> 
> So for updating the logic of `_num` functions, it might be easier to treat the case when neither of the inputs are NaNs first. Then you can deal with all the NaN cases afterward.
> 
> And `FPBits::is_nan()` function will return true for both quiet and signaling NaNs.

ok Ill try to address all of your review  in the next two hours.

lntue replied:
Just take your time, don't need to rush.  And thanks for bearing with me all the time!

Jobhdez replied:
No problem. I just finished addressing your review! I’m looking forward to merge this! Thanks

lntue commented on libc/test/src/math/smoke/FMinimumNumTest.h:
```diff
@@ -0,0 +1,100 @@
+    EXPECT_FP_EQ(FPBits::quiet_nan().get_val(), func(aNaN, aNaN));
+    EXPECT_FP_EQ(0.0, func(FPBits::quiet_nan().get_val(), 0.0));
+    EXPECT_FP_EQ(-0.0, func(-0.0, FPBits::quiet_nan().get_val()));
+    EXPECT_FP_EQ(0.0, func(FPBits::signaling_nan().get_val(), 0.0));
```
Chang this to `EXPECT_FP_EQ_WITH_EXCEPTION`

lntue commented on libc/test/src/math/smoke/FMinimumNumTest.h:
```diff
@@ -0,0 +1,100 @@
+    EXPECT_FP_EQ(inf, func(FPBits::quiet_nan().get_val(), inf));
+    EXPECT_FP_EQ(inf, func(FPBits::signaling_nan().get_val(), inf));
+    EXPECT_FP_EQ(neg_inf, func(neg_inf, FPBits::quiet_nan().get_val()));
+    EXPECT_FP_EQ(neg_inf, func(neg_inf, FPBits::signaling_nan().get_val()));
```
Change this to `EXPECT_FP_EQ_WITH_EXCEPTION`

lntue commented on libc/test/src/math/smoke/FMinimumNumTest.h:
```diff
@@ -0,0 +1,100 @@
+    EXPECT_FP_EQ(inf, func(FPBits::signaling_nan().get_val(), inf));
+    EXPECT_FP_EQ(neg_inf, func(neg_inf, FPBits::quiet_nan().get_val()));
+    EXPECT_FP_EQ(neg_inf, func(neg_inf, FPBits::signaling_nan().get_val()));
+    EXPECT_FP_EQ(FPBits::quiet_nan().get_val(), func(aNaN, aNaN));
```
Change this to `EXPECT_EQ(FPBits(aNaN).uintval(), FPBits(func(aNaN, aNaN)).uintval());

lntue replied:
Change this to `EXPECT_EQ(FPBits(aNaN).uintval(), FPBits(func(aNaN, aNaN)).uintval());

lntue commented on libc/test/src/math/smoke/FMinimumNumTest.h:
```diff
@@ -0,0 +1,100 @@
+        func(aNaN, FPBits::signaling_nan().get_val()), FE_INVALID);
+    EXPECT_FP_IS_NAN_WITH_EXCEPTION(
+        func(FPBits::signaling_nan().get_val(), aNaN), FE_INVALID);
+  }
```
Can you also add the following tests:
```
EXPECT_EQ(FPBits(aNaN).uintval(), FPBits(func(aNaN, sNaN)).uintval());
EXPECT_EQ(FPBits(aNaN).uintval(), FPBits(func(sNaN, aNaN)).uintval());
EXPECT_EQ(FPBits(aNaN).uintval(), FPBits(func(sNaN, sNaN)).uintval());
```

Commit:
address review - refactor, fmake correct code

Commit:
format code

lntue commented on libc/src/__support/FPUtil/BasicOperations.h:
```diff
@@ -58,6 +59,126 @@ LIBC_INLINE T fmax(T x, T y) {
+LIBC_INLINE T fmaximum_num(T x, T y) {
+  FPBits<T> bitx(x), bity(y);
+
+  if (bitx.is_signaling_nan()) {
+    fputil::raise_except_if_required(FE_INVALID);
+    if (bity.is_nan())
+      return FPBits<T>::quiet_nan().get_val();
+    return y;
+  }
+  if (bity.is_signaling_nan()) {
+    fputil::raise_except_if_required(FE_INVALID);
+    if (bitx.is_nan())
+      return FPBits<T>::quiet_nan().get_val();
+    return x;
+  }
+  if (bitx.is_quiet_nan())
+    return y;
+  if (bity.is_quiet_nan())
+    return x;
```
Sorry for the last request: this could be simplify a bit to:
```
if (bitx.is_nan()) {
  if (bitx.is_signaling_nan())
    fputil::raise_except_if_required(FE_INVALID)
  if (bity.is_nan())
    return FPBits<T>::quiet_nan().get_val();
  return y;
}
if (bity.is_nan()) {
  if (bity.is_signaling_nan())
    fputil::raise_except_if_required(FE_INVALID)
  // x is not NaN.
  return x;
}
```

The same for `fminimum_num`.

Jobhdez replied:
I used this code buts now the NaN tests are failing. Do i go back to what I had it?

lntue replied:
Sorry the logic I posted above is not correct for `fmaximum_num(aNaN, sNaN)`.  The corrected version is:
```
  if (bitx.is_signaling_nan() || bity.is_signaling_nan()) {
    fputil::raise_except_if_required(FE_INVALID);
    if (bitx.is_nan() && bity.is_nan())
      return FPBits<T>::quiet_nan().get_val();
  }
  if (bitx.is_nan())
    return y;
  if (bity.is_nan())
    return x;
  if (bitx.sign() != bity.sign())
    return (bitx.is_neg() ? y : x);
  return x > y ? x : y;
```

`fminimum_num` can be updated similarly.

Commit:
address review

Commit:
format code

Commit:
address review

lntue reviewed (approved):
Thanks for the patch!  Everything is clean now!

Jobhdez commented:
> Thanks for the patch! Everything is clean now!

Thank you!

arsenm commented:
> The GPU should use `__builtin_fmin` and `__builtin_fmax`. 

No, these are different operations. Not sure if we have a wired up builtin for these
